src/stores/chat.ts

```diff
@@ -0,0 +1,162 @@
+
+type ProcessChatBlocksOptions = {
+	/** if given, look for Set that has it */
+	mergeBy?: number
+}
+
+/**
+ * Check, if two sets intersect with each other
+ * Same complexity and result as !Set.prototype.isDisjointFrom()
+ */
+function checkIfIntersect(parentBlock: Set<number>, childBlock: Set<number>): boolean {
+	for (const id of childBlock) {
+		if (parentBlock.has(id)) {
+			return true
+		}
+	}
+	return false
+}
+
+/**
+ * Store for conversation chats
+ */
+export const useChatStore = defineStore('chat', () => {
+	const store = useStore()
+
+	const chatBlocks = reactive<TokenMap<Set<number>[]>>({})
+
+	/**
+	 * Returns list of messages, belonging to current context
+	 */
+	function getMessagesList(token: string): ChatMessage[] {
+		if (!store.state.messagesStore.messages[token] || !chatBlocks[token]) {
+			return []
+		}
+
+		// FIXME temporary show all messages from al blocks - no behaviour change
+		return Array.from(chatBlocks[token].flatMap((set) => Array.from(set)))
+			.sort((a, b) => a - b)
+			.reduce<ChatMessage[]>((acc, id) => {
+				const message = store.state.messagesStore.messages[token][id]
+				if (message) {
+					// If message is not found in the store, it's an invisible system or expired message
+					acc.push(message)
+				}
+				return acc
+			}, [])
+	}
+
+	/**
+	 * Populate chat blocks from given arrays of messages
+	 * If blocks already exist, try to extend them
+	 */
+	function processChatBlocks(token: string, messages: ChatMessage[], options?: ProcessChatBlocksOptions): void {
+		const newMessageIds = messages.map((message) => message.id)
+		const newMessageIdsSet = new Set(newMessageIds)
+
+		if (!chatBlocks[token]) {
+			// If no blocks exist, create a new one with the first message. First in array will be considered main block
+			chatBlocks[token] = [newMessageIdsSet]
+			return
+		}
+
+		if (options?.mergeBy) {
+			newMessageIdsSet.add(options.mergeBy)
+		}
+
+		if (!chatBlocks[token].some((parentBlock) => tryMergeBlocks(parentBlock, newMessageIdsSet))) {
+			chatBlocks[token].push(newMessageIdsSet)
+		}
+
+		mergeAndSortChatBlocks(token)
+	}
+
+	/**
+	 * Check, if extra blocks are intersecting with main, and merge them in this case
+	 */
+	function mergeAndSortChatBlocks(token: string) {
+		chatBlocks[token] = chatBlocks[token]
+			.reduce<Set<number>[]>((acc, block, index) => {
+				if (acc.length === 0 || !tryMergeBlocks(acc.at(-1)!, block)) {
+					acc.push(block)
+				}
+				return acc
+			}, [])
+			// Sort blocks in descending order by max id. First in array will be considered main block (recent messages)
+			.sort((a, b) => Math.max(...Array.from(b)) - Math.max(...Array.from(a)))
+	}
+
+	/**
+	 * Check, if child block is intersecting with parent, and extend parent in this case
+	 * Returns true if parent was extended, false otherwise
+	 */
+	function tryMergeBlocks(parentBlock: Set<number>, childBlock: Set<number>): boolean {
+		if (checkIfIntersect(parentBlock, childBlock)) {
```

Comment:
```suggestion
		if (Array.from(childBlock).some((id) => parentBlock.has(id))) 
```

~~checkIfIntersect~~

Reply:
Is it different from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/isDisjointFrom ?

Reply:
Both are `O(n)` we merge `k+n` and `k*n` to `n` afaik

Reply:
Still a bit more performant to iterate Set, rather than convert Set to Array and iterate Array

Reply:
> It's the same, but Set methods have some issues with node/typescript, haven't figured yet. Possible because of Node 20, but also won't work on Safari < 17

Other new ES features are also not available in older browsers. That's covered by transpiler and polyfills

Reply:
Then I will check if fixable for tsc and jest